lib/ts/recipe/thirdparty/types.ts

```diff
@@ -98,21 +98,21 @@ export type TypeProvider = {
 
     getConfigForClientType: (input: {
         clientType?: string;
-        userContext: UserContext;
+        userContext?: UserContext;
     }) => Promise<ProviderConfigForClientType>;
     getAuthorisationRedirectURL: (input: {
         redirectURIOnProviderDashboard: string;
-        userContext: UserContext;
+        userContext?: UserContext;
```

Comment:
we do nto want to make these optional since they are used internally as well. It would be better to expose UserContext instead so that devs can import it and create that and pass it in. These APIs are not used that often by users anyway

Reply:
Thanks for the quick response. 🙏🏾 Some thoughts:

> It would be better to expose UserContext instead so that devs can import it and create that and pass it in.

`UserContext` is just a type, right? Not a `class`? Are you saying that developers should do this instead?

```ts
import type { UserContext } from "supertokens-node/types";

// ...
const { urlWithQueryParams, pkceCodeVerifier } = await provider.getAuthorisationRedirectURL({
  userContext: undefined as UserContext, // or `{}`
  redirectURIOnProviderDashboard: createReturnURLForProvider(providerId, returnUrl),
});
```

(If yes, I think this is already doable today.)

---

> These APIs are not used that often by users anyway

This might be because users aren't aware of how to leverage the Node SDK effectively yet. As more examples come out documenting how to leverage this tool, that might change. I believe this is at least true for [SvelteKit](https://github.com/ITenthusiasm/svelte-kit-supertokens/issues/4).

Even so, as things stand, developers who want to implement their own API routes for the `ThirdParty` recipe are forced to use the [publicly-documented](https://supertokens.com/docs/nodejs/modules/recipe_thirdparty.html#TypeProvider) `TypeProvider` methods. And those methods currently require the `UserContext` argument even though it technically isn't required on the developer's end. This creates a slightly confusing maintenance experience for devs and introduces extra LOC unnecessarily.

One time when I was working on my `@form-observer/core` package, I was asking a friend about a certain API design. I was considering a design that would create a slight inconvenience to developers, but would slightly make maintenance easier on my end. What he had to say stuck out to me:

> Well I don't think the topic warrants dictating user code patterns based on library code's needs. If it were something that would affect performance if it were done in X way and not Y way, then yeah. Otherwise make it work...

You don't have to agree with that philosophy, and as the library maintainer you make the decisions for the project. But I do think it's worth considering prioritizing UX.

---

Is there a way to indicate to developers that `userContext` is optional, while also guarding the project's code in such a way that `userContext` is always passed in?

If not, another alternative is to make `userContext: UserContext | undefined`. This is similar to [`ThirdParty.getProvider`](https://supertokens.com/docs/nodejs/modules/recipe_thirdparty.html#getProvider-1): It makes `clientType: string | undefined`, which I'm assuming is an intentional design decision to make a distinction between a purposefully-excluded `clientType` and an explicitly-included `string`.

This would be an option that "splits the inconvenience" down the middle: Maintainers would technically be able to use `undefined` (bad) but would still be forced to provide the `userContext` (good), and users would still have to specify the property (bad) but without doing type casting on `undefined` (good).

---

Sorry I know that's a mouthful. Interested in hearing your thoughts.

Reply:
I think we can expose the type, and then ask users to pass in `{} as UserContext` when using the functions.

Reply:
I don't think passing `{} as UserContext` would be a fair thing to require of devs because it would force a performance cost on them. Specifically, devs would be forced to create throwaway memory (which also impacts garbage collection) on every request related to `ThirdParty` authentication. (Throwaway memory will be created both when a user tries to authenticate with a provider, and when they're redirected back from the provider's domain.)

It may be a small cost, but JS apps are often bloated and need as many performance gains as they can get. A developer recently released a video talking about the significant memory costs that come from using Next.js alone, and how that impacted server hosting costs. I don't think a library should push devs to add to that cost.

If the intent was to push responsibility to developers, `undefined as UserContext` would be more appropriate (but still an odd UX).

If possible, I'd still like to hear about any potential alternative solutions, or your thoughts on the middleground I proposed.

Reply:
I see. We can do something like `undefined as UserContext`

Reply:
Understood. I would still like to hear any thoughts or comments you have about other alternatives. Although what you're describing is an option, I still think it's a worst case scenario for developers. And given comments like [this one](https://discord.com/channels/603466164219281420/1294311977228701807/1295772655462645831), I think there's a realistic possibility of more people using this API as they become aware of using it as an option.

Is there any reason why `userContext: UserContext | undefined` would be unacceptable? You said that your concern was to remember to pass `userContext`, and that type will still force you to pass a value to that property.

Reply:
Update:

I thought `undefined as UserContext` would work because it seemed to be fine in the `supertokens-node` codebase. But for applications using more strict `tsconfig` configurations (which are more common), this approach does not work. This would mean that `as any` has to be used; but this is forbidden in many TS ESLint configurations.

This again makes me think that the cleanest solution is to change the type of `userContext` itself.

Reply:
The only issue with making `UserContext = Branded<Record<string, any>, "UserContext"> | undefined` is that it makes it possible for devs of the sdk to pass in undefined for a userContext when calling internal functions, thereby loosing it. I just dont want to have to add another point to manually check during reviews.

Reply:
I would instead make the provider functions accept an optional userContext in this case, so users just dont need to pass anything. Sure, this way our devs who maintain the sdk can forget to pass in the usercontext to the provider functions, but they are used in very limited places in the sdk anyway, so not too much to check.

Reply:
Okay. I thoroughly appreciate your openness to the idea of making the property optional. 🙇🏾‍♂️ In that case, would you be able to remove the Change Request? Or are there other changes that you'd like me to make?